sdks/go/pkg/beam/io/mongodbio/read.go

```diff
@@ -0,0 +1,469 @@
+	if err != nil {
+		return nil, fmt.Errorf("error executing bucketAuto aggregation: %w", err)
+	}
+
+	var buckets []bucket
+	if err = cursor.All(ctx, &buckets); err != nil {
+		return nil, fmt.Errorf("error decoding buckets: %w", err)
+	}
+
+	return buckets, nil
+}
+
+func idFiltersFromBuckets(buckets []bucket) []bson.M {
+	idFilters := make([]bson.M, len(buckets))
+
+	for i := 0; i < len(buckets); i++ {
+		filter := bson.M{}
+
+		if i != 0 {
+			filter["$gt"] = buckets[i].ID.Min
+		}
+
+		if i != len(buckets)-1 {
+			filter["$lte"] = buckets[i].ID.Max
+		}
+
+		if len(filter) == 0 {
+			idFilters[i] = filter
+		} else {
+			idFilters[i] = bson.M{"_id": filter}
+		}
+	}
+
+	return idFilters
+}
+
+type splitVectorFn struct {
+	mongoDBFn
+	BundleSize int64
+}
+
+func newSplitVectorFn(
+	uri string,
+	database string,
+	collection string,
+	option *ReadOption,
+) *splitVectorFn {
+	return &splitVectorFn{
+		mongoDBFn: mongoDBFn{
+			URI:        uri,
+			Database:   database,
+			Collection: collection,
+		},
+		BundleSize: option.BundleSize,
+	}
+}
+
+func (fn *splitVectorFn) ProcessElement(
+	ctx context.Context,
+	_ []byte,
+	emit func(bson.M),
+) error {
+	chunkSize := getChunkSize(fn.BundleSize)
+
+	splitKeys, err := fn.getSplitKeys(ctx, chunkSize)
+	if err != nil {
+		return err
+	}
+
+	idFilters := idFiltersFromSplits(splitKeys)
+
+	for _, filter := range idFilters {
+		emit(filter)
+	}
+
+	return nil
+}
+
+func getChunkSize(bundleSize int64) int64 {
+	var chunkSize int64
+
+	if bundleSize < minSplitVectorChunkSize {
+		chunkSize = minSplitVectorChunkSize
+	} else if bundleSize > maxSplitVectorChunkSize {
+		chunkSize = maxSplitVectorChunkSize
+	} else {
+		chunkSize = bundleSize
+	}
+
+	return chunkSize
+}
+
+type splitVector struct {
+	SplitKeys []splitKey `bson:"splitKeys"`
+}
+
+type splitKey struct {
+	ID any `bson:"_id"`
+}
+
+func (fn *splitVectorFn) getSplitKeys(ctx context.Context, chunkSize int64) ([]splitKey, error) {
+	cmd := bson.D{
+		{Key: "splitVector", Value: fmt.Sprintf("%s.%s", fn.Database, fn.Collection)},
+		{Key: "keyPattern", Value: bson.D{{Key: "_id", Value: 1}}},
+		{Key: "maxChunkSizeBytes", Value: chunkSize},
+	}
+
+	opts := options.RunCmd().SetReadPreference(readpref.Primary())
+
+	var vector splitVector
+	if err := fn.collection.Database().RunCommand(ctx, cmd, opts).Decode(&vector); err != nil {
+		return nil, fmt.Errorf("error executing splitVector command: %w", err)
+	}
+
+	return vector.SplitKeys, nil
+}
+
+func idFiltersFromSplits(splitKeys []splitKey) []bson.M {
+	idFilters := make([]bson.M, len(splitKeys)+1)
+
+	for i := 0; i < len(splitKeys)+1; i++ {
+		filter := bson.M{}
+
+		if i > 0 {
+			filter["$gt"] = splitKeys[i-1].ID
+		}
+
+		if i < len(splitKeys) {
+			filter["$lte"] = splitKeys[i].ID
+		}
+
+		if len(filter) == 0 {
+			idFilters[i] = filter
+		} else {
+			idFilters[i] = bson.M{"_id": filter}
+		}
+	}
+
+	return idFilters
+}
+
+type readFn struct {
+	mongoDBFn
+	Filter []byte
```

Comment:
I can have DoFn `ProcessElement` methods with elements of type `bson.M` (bucketAutoFn, splitVectorFn --> readFn), however if defining the `Filter` field of `readFn` as type bson.M (which is a `map[string]interface{}`) the program panics due to being unable to convert unsupported type `interface{}` to schema. Is there anything I can do to enable having the Filter as bson.M or will I have to go with this workaround of using a `[]byte` which is decoded into a bson.M within the ProcessElement method?

Reply:
Unfortunately, not. This is the current correct/best way. `interface{}/any` are unencodable by nature, and at decode time, concrete runtime types would be lost if they were simply JSON or BSON encoded by the framework. The framework works best with concrete types.

Ideally it should do the decoding in a `Setup()` method instead of `ProcessElement()`, and store the values in an unexported field of type `bson.M`. Setup will be called once during the instance's lifetime, while ProcessElement is called for every element. Though in practice, since these are kicked off with `beam.Impulse()`, those are the same.

Where the `Setup` approach shows value is when the connection/state information for the reads (or the filters, etc) is passed in as elements into the transform, rather than as static configuration. That way, for example, the results from several queries can be emitted from the single transform. This would enable longer lived clients and so on.

Certainly not required, and I'm not familiar enough to make the design decision. It's possible to migrate to that at a later moment anyway, if it seems useful, or simply have a different entry point.

Reply:
Good point, updated